File: Classes/NKJMultiMovieCaptureView.m
//
//  NKJMultiMovieCaptureView.m
//
//  Created by nakajijapan.
//  Copyright 2014 nakajijapan. All rights reserved.
//

#import "NKJMultiMovieCaptureView.h"
#import <AVFoundation/AVFoundation.h>
#import <AudioToolbox/AudioToolbox.h>



@interface NKJMultiMovieCaptureView () <AVCaptureVideoDataOutputSampleBufferDelegate, AVCaptureAudioDataOutputSampleBufferDelegate>

@property BOOL touching;
@property AVCaptureSession *captureSession;
@property AVCaptureVideoPreviewLayer *previewLayer;
@property AVAssetWriter *assetWriter;
@property AVAssetWriterInput *assetWriterInputVideo;
@property AVAssetWriterInput *assetWriterInputAudio;
@property AVCaptureVideoDataOutput *captureVideoOutput;
@property AVCaptureAudioDataOutput *captureAudioOutput;

@property NSMutableDictionary *videoSettings;
@property NSMutableDictionary *audioSettings;


@property CMTime recordStartTime;
@property NSURL *outputURL;
@property dispatch_queue_t movieWritingQueue;

@end

@implementation NKJMultiMovieCaptureView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor        = [UIColor blackColor];
        self.userInteractionEnabled = YES;
        self.movieURLs              = [NSMutableArray array];
        self.touching               = NO;
        self.recordStartTime        = kCMTimeZero;

        // device
        AVCaptureDevice* device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];

        // Video
        AVCaptureDeviceInput* videoInput = [AVCaptureDeviceInput deviceInputWithDevice:device error:NULL];
        AVCaptureVideoDataOutput* videoOutput = [[AVCaptureVideoDataOutput alloc] init];
        [videoOutput setSampleBufferDelegate:self queue:dispatch_get_main_queue()];
        [videoOutput setVideoSettings: [NSDictionary dictionaryWithObjectsAndKeys:
                                       [NSNumber numberWithInt:kCVPixelFormatType_32BGRA], (id)kCVPixelBufferPixelFormatTypeKey,
                                       nil]];


        // Audio
        NSError *error = nil;
        AVCaptureDevice *audioCaptureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeAudio];
        AVCaptureDeviceInput *audioInput = [AVCaptureDeviceInput deviceInputWithDevice:audioCaptureDevice error:&error];
        AVCaptureAudioDataOutput *audioOutput = [[AVCaptureAudioDataOutput alloc] init];
        [audioOutput setSampleBufferDelegate:self queue:dispatch_get_main_queue()];


        // Setting Camera
        AVCaptureConnection *videoConnection = NULL;
        [self.captureSession beginConfiguration];

        for ( AVCaptureConnection *connection in [videoOutput connections] ) {
            for ( AVCaptureInputPort *port in [connection inputPorts] ) {

                NSLog(@"%@", port);
                if ( [[port mediaType] isEqual:AVMediaTypeVideo] ) {
                    videoConnection = connection;
                }
            }
        }

        // portrait orientation
        [videoConnection setVideoOrientation:AVCaptureVideoOrientationPortrait];

        // initialize capture session
        self.captureSession = [[AVCaptureSession alloc] init];
        if ([self.captureSession canAddInput:videoInput]) {
            [self.captureSession addInput:videoInput];
        }
        if ([self.captureSession canAddOutput:videoOutput]) {
            [self.captureSession addOutput:videoOutput];
        }
        if ([self.captureSession canAddInput:audioInput]) {
            [self.captureSession addInput:audioInput];
        }
        if ([self.captureSession canAddOutput:audioOutput]) {
            [self.captureSession addOutput:audioOutput];
        }
        [self.captureSession setSessionPreset: AVCaptureSessionPreset1280x720];


        // PreviewLayer
        self.previewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:self.captureSession];

        // aspect
        [self.previewLayer setVideoGravity:AVLayerVideoGravityResizeAspectFill];

        // display size
        CALayer *rootLayer = [self layer];
        [rootLayer setMasksToBounds:YES];
        [self.previewLayer setFrame:CGRectMake(0, 0, frame.size.width, frame.size.height)];
        [rootLayer addSublayer:self.previewLayer];

        // reflection a changes
        [self.captureSession commitConfiguration];

        // session start
        [self.captureSession startRunning];


        // video setting
        self.videoSettings = [NSMutableDictionary dictionaryWithObjectsAndKeys:
                              AVVideoCodecH264, AVVideoCodecKey,
                              @1280, AVVideoWidthKey,
                              @720,  AVVideoHeightKey,
                              [NSDictionary dictionaryWithObjectsAndKeys:
                               @30, AVVideoMaxKeyFrameIntervalKey,
                               nil], AVVideoCompressionPropertiesKey,
                              nil];

        // audio setting
        self.audioSettings = [NSMutableDictionary dictionaryWithObjectsAndKeys:
                              @44100.0, AVSampleRateKey,
                              @2,       AVNumberOfChannelsKey,
                              @16,      AVLinearPCMBitDepthKey,
                              [NSNumber numberWithInt: kAudioFormatLinearPCM], AVFormatIDKey,
                              [NSNumber numberWithBool:NO], AVLinearPCMIsFloatKey,
                              [NSNumber numberWithBool:NO], AVLinearPCMIsBigEndianKey,
                              [NSNumber numberWithBool:NO], AVLinearPCMIsNonInterleaved,
                              [NSData data], AVChannelLayoutKey,
                              nil];
    }
    return self;
}

- (void)setVideoSettingWithDictionary:(NSDictionary *)settings
{
    for (id keyName in [settings allKeys]) {
        [self.videoSettings setObject:[settings objectForKey:keyName] forKey:keyName];
    }
}

- (void)setAudioSettingWithDictionary:(NSDictionary *)settings
{
    for (id keyName in [settings allKeys]) {
        [self.audioSettings setObject:[settings objectForKey:keyName] forKey:keyName];
    }
}

#pragma mark - AVCaptureVideoDataOutputSampleBufferDelegate, AVCaptureAudioDataOutputSampleBufferDelegate

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
{

    if (!CMSampleBufferDataIsReady(sampleBuffer)) {
        NSLog(@"sampleBuffer data is not ready");
    }

    CMTime currentTime = CMSampleBufferGetPresentationTimeStamp(sampleBuffer);

    if (self.touching) {

        // Video
        if ([captureOutput isKindOfClass:[AVCaptureVideoDataOutput class]]) {
            [self addSamplebuffer:sampleBuffer withWriterInput:(AVAssetWriterInput *)self.assetWriterInputVideo];
        }
        // Audio
        else if ([captureOutput isKindOfClass:[AVCaptureAudioDataOutput class]]) {
            [self addSamplebuffer:sampleBuffer withWriterInput:(AVAssetWriterInput *)self.assetWriterInputAudio];
        }

    }

    self.recordStartTime = currentTime;
}

- (void)addSamplebuffer:(CMSampleBufferRef)sampleBuffer withWriterInput:(AVAssetWriterInput *)assetWriterInput
{
    CMFormatDescriptionRef formatDescription = CMSampleBufferGetFormatDescription(sampleBuffer);

    CFRetain(sampleBuffer);
    CFRetain(formatDescription);
    dispatch_async(self.movieWritingQueue, ^{
        if (![assetWriterInput isReadyForMoreMediaData]) {
            NSLog(@"Not ready for data :(");
        }
        //NSLog(@"Trying to append");

        if (_assetWriter.status == AVAssetWriterStatusUnknown) {
            NSLog(@"AVAssetWriterStatusUnknown");
        }
        if (_assetWriter.status == AVAssetWriterStatusWriting) {
            //NSLog(@"AVAssetWriterStatusWriting");

            if (assetWriterInput.readyForMoreMediaData) {

                if (![assetWriterInput appendSampleBuffer:sampleBuffer]) {
                    NSLog(@"%@",[self.assetWriter error]);
                }

            }
        }
        else if (_assetWriter.status == AVAssetWriterStatusFailed) {
            NSLog(@"AVAssetWriterStatusFailed");
            NSLog(@"%@",[self.assetWriter error]);
        }
        else if (_assetWriter.status == AVAssetWriterStatusCancelled) {
            NSLog(@"AVAssetWriterStatusCancelled");
        }
        else if (_assetWriter.status == AVAssetWriterStatusCompleted) {
            NSLog(@"AVAssetWriterStatusCompleted");
        }

        CFRelease(sampleBuffer);
        CFRelease(formatDescription);
    });
}

#pragma mark - Touch Events

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    self.touching = YES;

    // Output
    NSError *error;
    NSString *fileName   = [NSString stringWithFormat:@"output%02d.mov", (int)[self.movieURLs count] + 1];
    NSString *outputPath = [[NSString alloc] initWithFormat:@"%@%@", NSTemporaryDirectory(), fileName];
    self.outputURL       = [[NSURL alloc] initFileURLWithPath:outputPath];
    NSLog(@"outputPath = %@", outputPath);

    // delete file before save the one
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if ([fileManager fileExistsAtPath:outputPath]) {

        if ([fileManager removeItemAtPath:outputPath error:&error] == NO) {
            //NSLog(@"failed deleting file");
        }

    }

    // AVAssetWriter
    self.assetWriter = [AVAssetWriter assetWriterWithURL:self.outputURL fileType:AVFileTypeQuickTimeMovie error:&error];
    if (error != nil) {
        //NSLog(@"Creation of assetWriter resulting in a non-nil error");
    }

    // movie
    NSDictionary *videoSetting = self.videoSettings;
    self.assetWriterInputVideo = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:videoSetting];
    [self.assetWriterInputVideo setExpectsMediaDataInRealTime:YES];
    if (self.assetWriterInputVideo == nil) {
        NSLog(@"assetWriterInput is nil");
    }

    // audio
    NSDictionary *audioSetting = self.audioSettings;
    self.assetWriterInputAudio = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeAudio outputSettings:audioSetting];
    [self.assetWriterInputAudio setExpectsMediaDataInRealTime:YES];
    if (self.assetWriterInputAudio == nil) {
        NSLog(@"assetWriterInputAudio is nil");
    }

    [self.assetWriter addInput:self.assetWriterInputVideo];
    [self.assetWriter addInput:self.assetWriterInputAudio];

    // queue
    self.movieWritingQueue = dispatch_queue_create("Movie Writing Queue", DISPATCH_QUEUE_SERIAL);

    // Record
    NSLog(@"[Starting to record]");
    dispatch_async(self.movieWritingQueue, ^{
        [self.assetWriter startWriting];
        [self.assetWriter startSessionAtSourceTime:self.recordStartTime];
    });

}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    self.touching = NO;

    NSLog(@"[Stopping recording] duration : %f", CMTimeGetSeconds(self.recordStartTime));

    [self.assetWriterInputVideo markAsFinished];
    [self.assetWriterInputAudio markAsFinished];
    [self.assetWriter endSessionAtSourceTime:self.recordStartTime];
    [self.assetWriter finishWritingWithCompletionHandler:^{
        NSLog(@"self.assetWriter finishWritingWithCompletionHandler");

        [self.movieURLs addObject:self.outputURL];
        NSLog(@"%@", self.movieURLs);

    }];

    NSLog(@"Export done");
}
@end

File: Pod/Classes/NKJMultiMovieCaptureView.swift
//
//  NKJMultiMovieCaptureView.swift
//  Pods
//
//  Created by nakajijapan on 6/19/15.
//
//

import UIKit
import CoreFoundation
import AVFoundation
import AudioToolbox

public class NKJMultiMovieCaptureView: UIView, AVCaptureVideoDataOutputSampleBufferDelegate, AVCaptureAudioDataOutputSampleBufferDelegate {

    public var movieURLs:[NSURL] = []
    
    var touching:Bool = false
    var captureSession = AVCaptureSession()
    var previewLayer:AVCaptureVideoPreviewLayer!
    var assetWriter:AVAssetWriter?
    var assetWriterInputVideo:AVAssetWriterInput!
    var assetWriterInputAudio:AVAssetWriterInput!
    var captureVideoOutput:AVCaptureVideoDataOutput!
    var captureAudioOutput:AVCaptureAudioDataOutput!
    
    var videoSettings = [NSObject : AnyObject]()
    var audioSettings = [NSObject : AnyObject]()
    
    var recordStartTime = kCMTimeZero
    var outputURL:NSURL?
    var movieWritingQueue:dispatch_queue_t?

    override public init(frame: CGRect) {
        super.init(frame: frame)
        
        self.userInteractionEnabled = true
        self.backgroundColor = UIColor.blackColor()
        
        // device
        let device = AVCaptureDevice.defaultDeviceWithMediaType(AVMediaTypeVideo)
        
        // video
        var videoInput: AVCaptureDeviceInput!
        do {
            videoInput = try AVCaptureDeviceInput(device: device)
        } catch var error as NSError {
            videoInput = nil
            fatalError("error: \(error!.localizedDescription)")
        }

        var videoOutput = AVCaptureVideoDataOutput()
        videoOutput.setSampleBufferDelegate(self, queue: dispatch_get_main_queue())
        videoOutput.videoSettings = [kCVPixelBufferPixelFormatTypeKey: kCVPixelFormatType_32BGRA]

        // audio
        var audioCaptureDevice = AVCaptureDevice.defaultDeviceWithMediaType(AVMediaTypeAudio)
        var audioInput: AVCaptureDeviceInput!
        do {
            audioInput = try AVCaptureDeviceInput(device: audioCaptureDevice)
        } catch var error as NSError {
            audioInput = nil
            fatalError("error: \(error!.localizedDescription)")
        }

        var audioOutput = AVCaptureAudioDataOutput()
        audioOutput.setSampleBufferDelegate(self, queue: dispatch_get_main_queue())

        // setting camera
        var videoConnection:AVCaptureConnection? = nil
        self.captureSession.beginConfiguration()
        
        for connection in videoOutput.connections {
            
            if let captureConnection = connection as? AVCaptureConnection {

                for port in captureConnection.inputPorts {
                    
                    print("\(port)")
                    if let inputPort = port as? AVCaptureInputPort {
                        
                        if inputPort.mediaType == AVMediaTypeVideo {
                            videoConnection = captureConnection
                        }
                    }
                }
                
            }
            
        }
        
        // portrait orientation
        videoConnection?.videoOrientation = AVCaptureVideoOrientation.Portrait
        
        // initialize capture session
        if self.captureSession.canAddInput(videoInput) {
            self.captureSession.addInput(videoInput)
        }
        
        if self.captureSession.canAddOutput(videoOutput) {
            self.captureSession.addOutput(videoOutput)
        }
        
        if self.captureSession.canAddInput(audioInput) {
            self.captureSession.addInput(audioInput)
        }
        
        if self.captureSession.canAddOutput(audioOutput) {
            self.captureSession.addOutput(audioOutput)
        }
        self.captureSession.sessionPreset = AVCaptureSessionPreset1280x720
        
        // PreviewLayer
        self.previewLayer = AVCaptureVideoPreviewLayer(session: self.captureSession)
        
        // aspect
        self.previewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill
        
        // display size
        var rootLayer = self.layer
        rootLayer.masksToBounds = true
        self.previewLayer.frame = CGRectMake(0, 0, frame.size.width, frame.size.height)
        rootLayer.addSublayer(self.previewLayer)
        
        // reflection a changes
        self.captureSession.commitConfiguration()
        
        // session start
        self.captureSession.startRunning()
        
        // video setting
        self.videoSettings = [
            AVVideoCodecKey: AVVideoCodecH264,
            AVVideoWidthKey: 1280,
            AVVideoHeightKey: 720,
            AVVideoCompressionPropertiesKey: [AVVideoMaxKeyFrameIntervalKey: 30]
        ]
        
        self.audioSettings = [
            AVSampleRateKey: 44100.0,
            AVNumberOfChannelsKey: 2,
            AVLinearPCMBitDepthKey: 16,
            AVFormatIDKey: kAudioFormatLinearPCM,
            AVLinearPCMIsFloatKey: false,
            AVLinearPCMIsBigEndianKey: false,
            AVLinearPCMIsNonInterleaved: false,
            AVChannelLayoutKey: NSData()
        ]
        
    }

    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Public Method
    
    public func setVideoSettingWithDictionary(settings: Dictionary<NSObject, AnyObject>) {
        for keyName in settings.keys {
            self.videoSettings[keyName] = settings[keyName]
        }
    }

    public func setAudioSettingWithDictionary(settings: Dictionary<NSObject, AnyObject>) {
        for keyName in settings.keys {
            self.audioSettings[keyName] = settings[keyName]
        }
    }
    
    // MARK: - AVCaptureVideoDataOutputSampleBufferDelegate, AVCaptureAudioDataOutputSampleBufferDelegate
    
    public func captureOutput(captureOutput: AVCaptureOutput!, didOutputSampleBuffer sampleBuffer: CMSampleBuffer!, fromConnection connection: AVCaptureConnection!) {
    
        if CMSampleBufferDataIsReady(sampleBuffer) == 0 {
            print("sampleBuffer data is not ready")
        }
        
        let currentTime = CMSampleBufferGetPresentationTimeStamp(sampleBuffer)
        
        if self.touching {

            if captureOutput.isKindOfClass(AVCaptureVideoDataOutput.self) { // video
                self.addSamplebuffer(sampleBuffer, assetWriterInput: self.assetWriterInputVideo as AVAssetWriterInput)
            } else if captureOutput.isKindOfClass(AVCaptureAudioDataOutput.self) { // audio
                self.addSamplebuffer(sampleBuffer, assetWriterInput: self.assetWriterInputAudio as AVAssetWriterInput)
            }
            
        }
        
        self.recordStartTime = currentTime;
    }
    
    func addSamplebuffer(sampleBuffer: CMSampleBufferRef, assetWriterInput: AVAssetWriterInput) {

        var formatDescription = CMSampleBufferGetFormatDescription(sampleBuffer)

        dispatch_async(self.movieWritingQueue!) { () -> Void in
            
            if !assetWriterInput.readyForMoreMediaData {
                print("Not ready for data")
            }
            
            if self.assetWriter?.status == AVAssetWriterStatus.Unknown {
                print("AVAssetWriterStatus.Unknown")
            }

            if self.assetWriter?.status == AVAssetWriterStatus.Writing {
                //println("AVAssetWriterStatus.Writing")
                
                if assetWriterInput.readyForMoreMediaData {

                    if !assetWriterInput.appendSampleBuffer(sampleBuffer) {
                        print("\(self.assetWriter?.error)");
                    }
                    
                }

            } else if self.assetWriter?.status == AVAssetWriterStatus.Failed {
                print("AVAssetWriterStatus.Failed")
            } else if self.assetWriter?.status == AVAssetWriterStatus.Cancelled {
                print("AVAssetWriterStatus.Cancelled")
            } else if self.assetWriter?.status == AVAssetWriterStatus.Completed {
                print("AVAssetWriterStatus.Completed")
            }
            
        }
        
    }

    // MARK: - Touch Events
    override public func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        
        self.touching = true
        
        let fileName = String(format: "output%02d.mov", arguments: [self.movieURLs.count + 1])
        let outputPath = NSTemporaryDirectory().stringByAppendingString(fileName)
        self.outputURL = NSURL(fileURLWithPath: outputPath)
        
        // delete file before save the one
        let fileManager = NSFileManager.defaultManager()
        if fileManager.fileExistsAtPath(outputPath) {
            
            do {
                try fileManager.removeItemAtPath(outputPath)
            } catch let error as NSError {
                print("failed deleting file (\(error.localizedDescription))")
            }
        }

        // AVAssetWriter
        do {

            self.assetWriter = try AVAssetWriter(URL: self.outputURL, fileType: AVFileTypeQuickTimeMovie)
        } catch let error as NSError {
            self.assetWriter = nil
            print("creation of assetWriter resulting in a non-nil error ((\(error.localizedDescription)))")
        }
        
        // movie
        self.assetWriterInputVideo = AVAssetWriterInput(mediaType: AVMediaTypeVideo, outputSettings: self.videoSettings)
        self.assetWriterInputVideo.expectsMediaDataInRealTime = true
        if self.assetWriterInputVideo == nil {
            print("assetWriterInputVideo is nil")
        }
        
        // audio
        self.assetWriterInputAudio = AVAssetWriterInput(mediaType: AVMediaTypeAudio, outputSettings: self.audioSettings)
        self.assetWriterInputAudio.expectsMediaDataInRealTime = true
        if self.assetWriterInputAudio == nil {
            print("assetWriterInputAudio is nil")
        }
        
        self.assetWriter?.addInput(self.assetWriterInputVideo)
        self.assetWriter?.addInput(self.assetWriterInputAudio)
        
        // queue
        self.movieWritingQueue = dispatch_queue_create("Movie Writing Queue", DISPATCH_QUEUE_SERIAL)

        // record
        dispatch_async(self.movieWritingQueue!, { () -> Void in
            self.assetWriter?.startWriting()
            self.assetWriter?.startSessionAtSourceTime(self.recordStartTime)
        })
    }
    
    public override func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent?) {
        
        self.touching = false
        
        //println("[stopping recording] duration :\(CMTimeGetSeconds(self.recordStartTime))")
        self.assetWriterInputVideo.markAsFinished()
        self.assetWriterInputAudio.markAsFinished()
        self.assetWriter?.endSessionAtSourceTime(self.recordStartTime)
        self.assetWriter?.finishWritingWithCompletionHandler({ () -> Void in

            //println("self.assetWriter finishWritingWithCompletionHandler")
            self.movieURLs.append(self.outputURL!)
            //println("\(self.outputURL)")
        })
        
    }

}
